File: Frontend_erp_cga/app/lib/formats.ts
/**
 * Formats de restitution — § 4 du dossier de design.
 *
 * Miroir de `Backend_erp_cga/app/shared/formats.py`. Les deux implémentations
 * doivent rester alignées : un montant rendu différemment côté serveur et côté
 * client dans un même écran est un défaut visible.
 *
 * Le séparateur de milliers est une espace insécable étroite (U+202F) : elle
 * empêche un montant de se couper en fin de ligne.
 */

export const ESPACE_FINE = " ";

const MOIS = [
  "janvier",
  "février",
  "mars",
  "avril",
  "mai",
  "juin",
  "juillet",
  "août",
  "septembre",
  "octobre",
  "novembre",
  "décembre",
] as const;

/**
 * Séparateur espace, aucune décimale, négatif entre parenthèses et jamais signé.
 * `2350000` → « 2 350 000 » · `-450000` → « (450 000) »
 */
export function montant(valeur: number | string): string {
  const nombre = Math.round(Number(valeur));
  if (!Number.isFinite(nombre)) return "—";
  const groupes = Math.abs(nombre)
    .toString()
    .replace(/\B(?=(\d{3})+(?!\d))/g, ESPACE_FINE);
  return nombre < 0 ? `(${groupes})` : groupes;
}

/** `2350000` → « 2 350 000 FCFA ». */
export function montantFcfa(valeur: number | string): string {
  return `${montant(valeur)}${ESPACE_FINE}FCFA`;
}

/** Virgule décimale, symbole séparé par une espace fine. `19.25` → « 19,25 % ». */
export function taux(valeur: number | string, decimales = 2): string {
  const nombre = Number(valeur);
  if (!Number.isFinite(nombre)) return "—";
  const rendu = nombre.toFixed(decimales).replace(/\.?0+$/, "");
  return `${rendu.replace(".", ",")}${ESPACE_FINE}%`;
}

/**
 * ⚠️ Pas 87 : les trois formats ci-dessous acceptent aussi une date-heure
 * (`2026-02-12T09:00:00`), dont ils ne gardent que le jour. Ils ajoutaient
 * `T00:00:00` à la chaîne reçue : une date-heure devenait `…T09:00:00T00:00:00`,
 * invalide, et l'écran affichait « le NaN/NaN/NaN » sous l'accusé d'un dépôt de TVA.
 * Corriger la fonction répare tous les appels, y compris ceux qui n'ont pas encore
 * rencontré de date-heure.
 */

/** Forme lisible. `2026-08-15` → « 15 août 2026 ». */
export function dateLongue(valeur: Date | string): string {
  const d = typeof valeur === "string" ? new Date(`${valeur.slice(0, 10)}T00:00:00`) : valeur;
  return `${d.getDate()} ${MOIS[d.getMonth()]} ${d.getFullYear()}`;
}

/** Forme compacte réservée aux tableaux denses. `2026-08-15` → « 15/08/2026 ». */
export function dateCourte(valeur: Date | string): string {
  const d = typeof valeur === "string" ? new Date(`${valeur.slice(0, 10)}T00:00:00`) : valeur;
  const jour = String(d.getDate()).padStart(2, "0");
  const mois = String(d.getMonth() + 1).padStart(2, "0");
  return `${jour}/${mois}/${d.getFullYear()}`;
}

/** Mois et année en toutes lettres. `2026-07-01` → « juillet 2026 ». */
export function periode(valeur: Date | string): string {
  const d = typeof valeur === "string" ? new Date(`${valeur.slice(0, 10)}T00:00:00`) : valeur;
  return `${MOIS[d.getMonth()]} ${d.getFullYear()}`;
}

File: Backend_erp_cga/app/partage/formats.py
"""Formats de restitution — § 4 du dossier de design.

Ces règles ne sont pas cosmétiques : une balance dont les montants ne sont pas alignés et
tabulaires est illisible, et un montant négatif rendu par un signe moins isolé se confond
avec un tiret dans une colonne dense.

Le séparateur de milliers est une **espace insécable étroite** (U+202F) : elle empêche un
montant de se couper en fin de ligne. Les tests de rendu la normalisent avant comparaison.
"""

from __future__ import annotations

from datetime import date, datetime
from decimal import ROUND_HALF_UP, Decimal

__all__ = [
    "ESPACE_FINE",
    "MOIS",
    "date_courte",
    "date_longue",
    "moment_long",
    "montant",
    "montant_fcfa",
    "taux",
]

#: Espace insécable étroite, U+202F.
ESPACE_FINE = " "

MOIS = (
    "janvier",
    "février",
    "mars",
    "avril",
    "mai",
    "juin",
    "juillet",
    "août",
    "septembre",
    "octobre",
    "novembre",
    "décembre",
)


def montant(valeur: Decimal | int | float) -> str:
    """Séparateur espace, aucune décimale, négatif entre parenthèses et jamais signé.

    2350000 → « 2 350 000 » · -450000 → « (450 000) »
    """
    entier = Decimal(str(valeur)).quantize(Decimal(1), rounding=ROUND_HALF_UP)
    groupes = f"{abs(entier):,}".replace(",", ESPACE_FINE)
    return f"({groupes})" if entier < 0 else groupes


def montant_fcfa(valeur: Decimal | int | float) -> str:
    """2350000 → « 2 350 000 FCFA »."""
    return f"{montant(valeur)}{ESPACE_FINE}FCFA"


def taux(valeur: Decimal | int | float, decimales: int = 2) -> str:
    """Virgule décimale, symbole séparé par une espace fine. 19.25 → « 19,25 % »."""
    rendu = f"{Decimal(str(valeur)):.{decimales}f}"
    if "." in rendu:
        rendu = rendu.rstrip("0").rstrip(".")
    return f"{rendu.replace('.', ',')}{ESPACE_FINE}%"


def date_longue(jour: date) -> str:
    """Forme lisible. 2026-08-15 → « 15 août 2026 »."""
    return f"{jour.day} {MOIS[jour.month - 1]} {jour.year}"


def date_courte(jour: date) -> str:
    """Forme compacte réservée aux tableaux denses. 2026-08-15 → « 15/08/2026 »."""
    return f"{jour.day:02d}/{jour.month:02d}/{jour.year}"


def moment_long(instant: datetime) -> str:
    """Un instant, lisible par un destinataire. « 30 août 2026 à 14h22 ».

    ⚠️ Réservé à ce qui est **lu par un humain** — un courriel, un écran. Une
    date d'expiration rendue en ISO (`2026-08-30T14:22:11`) oblige le lecteur à
    décoder un format machine pour savoir s'il lui reste une heure ou trois
    jours, et cette hésitation-là fait rater des délais.

    Sans fuseau affiché : le cabinet et ses adhérents sont tous en WAT, et
    ajouter « UTC+1 » sur chaque échéance ferait du bruit pour un lecteur qui
    n'a rien à convertir. ⚠️ Le jour où un adhérent de la diaspora existe, ce
    silence devient un défaut.
    """
    return f"{date_longue(instant.date())} à {instant.hour:02d}h{instant.minute:02d}"
